sdks/go/pkg/beam/core/runtime/exec/decode.go

```diff
@@ -0,0 +1,84 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exec
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/funcx"
+)
+
+//go:generate specialize --input=decoders.tmpl --x=data,universals
+
+// Decoder is a uniform custom encoder interface.
+type Decoder interface {
+	// Decode decodes the []byte in to a value of the given type.
+	Decode(reflect.Type, []byte) (reflect.Value, error)
+}
+
+var (
+	decoders   = make(map[string]func(reflect.Value) Decoder)
+	decodersMu sync.Mutex
+)
+
+// RegisterDecoder registers an custom decoder invoker for the given type,
+// such as "func(int)[]byte". If multiple decoder invokers are registered
+// for the same type, the last registration wins.
```

Comment:
Should we at least log a warning we're clobbering a registration? That would help a user in the case where they aren't getting the coder behavior they expect. Bonus points if we retain the filename and line of code that performed the registration.